lib/taskana-core/src/main/java/pro/taskana/impl/ClassificationServiceImpl.java

```diff
@@ -351,4 +352,78 @@ public Classification getClassificationByTask(Task task) throws ClassificationNo
+        if (existingTasksForClassificationInDomains) {
+            throw new ClassificationInUseException("There are Tasks which belong to this classification. Please complete them and try again.");
+        }
+
+        this.internalDeleteClassification(classificationKey, domains);
+    }
+
+    @Override
+    public void deleteRootClassification(String classificationKey) throws ClassificationInUseException, NotAuthorizedException {
+        String[] domains = classificationMapper.getDomainsForClassificationKey(classificationKey, CURRENT_CLASSIFICATIONS_VALID_UNTIL);
+
+        TaskServiceImpl taskService = (TaskServiceImpl) taskanaEngineImpl.getTaskService();
+        List<Task> tasks = new ArrayList<>();
+        tasks.addAll(taskService.createTaskQuery().classificationKeyIn(classificationKey).list());
+        if (!tasks.isEmpty()) {
+            throw new ClassificationInUseException("There are Tasks which belong to this classification. Please complete them and try again.");
+        }
+
+        String newParent = classificationMapper.findByKeyAndDomain(classificationKey, "", CURRENT_CLASSIFICATIONS_VALID_UNTIL).getParentClassificationKey();
+
+        try {
+            this.internalDeleteClassification(classificationKey, domains);
+        } catch (ClassificationNotFoundException e) {
+            // should not happen.
+            throw new InternalError("A internal error occured. There may be a problem with the 'getDomainsForClassificationKey' request of the classificationMapper.");
+        }
+
+        this.careForTheChildren(classificationKey, newParent);
+    }
+
+    private void internalDeleteClassification(String classificationKey, String... domains) throws ClassificationNotFoundException {
+        List<ClassificationImpl> deleteClassifications = new ArrayList<>();
+        for (String domain : domains) {
+            ClassificationImpl classification = classificationMapper.findByKeyAndDomain(classificationKey, domain, CURRENT_CLASSIFICATIONS_VALID_UNTIL);
+            if (classification == null) {
+                throw new ClassificationNotFoundException("there is no classification with the key " + classificationKey + " in the domain " + domain);
+            }
+            deleteClassifications.add(classification);
+        }
+        for (ClassificationImpl classification : deleteClassifications) {
+            classification.setValidUntil(Date.valueOf(LocalDate.now().minusDays(1)));
+            classificationMapper.update(classification);
+        }
+    }
+
+    /**
+     * This function is based on the assumption, that a domain classification never has another parent than its root classification.
+     * In this case the children of a deleted domain classification fall automatically back to the root classification.
+     */
+    private void careForTheChildren(String classificationKey, String newParent) throws NotAuthorizedException {
```

Comment:
There are to many call´s and bad readability. Please use caliic or longer names for lists/variables. Why are you removing a single root/parent-classification like this?

Reply:
Actually we better delete all children as well instead of manipulating the tree.